File: statelang/src/main/java/com/statelang/diagnostics/Report.java
package com.statelang.diagnostics;

import java.util.Collections;
import java.util.List;
import java.util.stream.Stream;

import javax.annotation.Nullable;

import com.statelang.tokenization.SourceLocation;
import com.statelang.tokenization.SourceSelection;
import com.statelang.tokenization.Token;

import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.experimental.Accessors;

@Accessors(fluent = true)
@AllArgsConstructor(access = AccessLevel.PRIVATE)
@Builder(toBuilder = true)
public final class Report {

	public static enum Severity {
		INFO,
		WARNING,
		ERROR,
	}

	public enum Kind {
		INVALID_TOKEN(Severity.ERROR),
		UNEXPECTED_TOKEN(Severity.ERROR),
		UNEXPECTED_END_OF_INPUT(Severity.ERROR),
		END_OF_INPUT_EXPECTED(Severity.ERROR),

		VALUE_EXPRESSION_EXPECTED(Severity.ERROR),
		CONDITION_EXPECTED(Severity.ERROR),

		AMBIGUOUS_DEFINITION(Severity.ERROR),
		MISSING_STATE_DEFINITION(Severity.ERROR),
		TOO_LITTLE_STATES(Severity.ERROR),

		DUPLICATE_IDENTIFIER(Severity.ERROR),

		UNDEFINED_OPERATOR(Severity.ERROR),
		UNDEFINED_VARIABLE(Severity.ERROR),
		UNDEFINED_STATE(Severity.ERROR),

		CONSTANT_ASSIGNMENT(Severity.ERROR),

		TYPE_ERROR(Severity.ERROR),

		UNREACHABLE_CODE(Severity.WARNING),
		UNREACHABLE_STATE(Severity.WARNING);

		@Getter
		private final Severity severity;

		private Kind(Severity severity) {
			this.severity = severity;
		}
	}

	@Getter
	private final SourceSelection selection;

	@Getter
	private final Kind kind;

	@Getter
	@Builder.Default
	private final Token.Kind unexpectedTokenKind = null;

	@Getter
	@Builder.Default
	private final List<Token.Kind> expectedTokenKinds = Collections.emptyList();

	@Getter
	@Nullable
	@Builder.Default
	private final String info = null;

	public Severity severity() {
		return kind.severity;
	}

	public SourceLocation location() {
		return selection.start();
	}

	@Override
	public String toString() {
		return severity() + " (" + selection + ") " + kind.name();
	}

	public static Report determineMostRelevant(Report first, Report second) {
		if (first.severity() != second.severity()) {
			return first.severity().ordinal() > second.severity().ordinal()
				? first
				: second;
		}

		if (first.selection().equals(second.selection())) {
			var firstExpectedTokens = first.expectedTokenKinds.stream();
			var secondExpectedTokens = second.expectedTokenKinds.stream();
			var expectedTokenKinds = Stream.concat(firstExpectedTokens, secondExpectedTokens).toList();

			return Report.builder()
				.selection(first.selection())
				.kind(first.kind())
				.expectedTokenKinds(expectedTokenKinds)
				.build();
		}

		if (first.location().equals(second.location())) {
			return first.selection().end().isBefore(second.selection().end())
				? first
				: second;
		}

		return first.location().isAfter(second.location())
			? first
			: second;
	}
}

File: stateviz/src/main/frontend/src/lib/statelang/report.ts
import * as monaco from 'monaco-editor';
import { joinWithLast } from '../utils';
import type { Report } from './model';

const REPORT_MESSAGE_MAP: Record<Report['kind'], (report: Report) => string> = {
	INVALID_TOKEN: report => 'invalid token',
	UNEXPECTED_TOKEN: report => 'unexpected token',
	UNEXPECTED_END_OF_INPUT: report => 'unexpected end of input',
	END_OF_INPUT_EXPECTED: report => 'end of input expected',
	VALUE_EXPRESSION_EXPECTED: report => 'value expression expected',
	CONDITION_EXPECTED: report => 'condition expected',
	VARIABLE_EXPECTED: report => 'variable expected',
	VARIABLE_OR_CONSTANT_EXPECTED: report => 'variable or constant expected',
	AMBIGUOUS_DEFINITION: report => 'ambiguous definition',
	MISSING_STATE_DEFINITION: report => 'missing state definition',
	TOO_LITTLE_STATES: report => 'too little states',
	DUPLICATE_IDENTIFIER: report => `duplicate identifier '${report.info}'`,
	UNDEFINED_OPERATOR: report => `operator ${report.info} is not defined`,
	UNDEFINED_VARIABLE: report => `variable ${report.info} is not defined`,
	UNDEFINED_STATE: report => `state ${report.info} is not defined`,
	CONSTANT_ASSIGNMENT: report => 'assignment to constant',
	TYPE_ERROR: report => {
		const [actualType, expectedType] = (report.info ?? '? ?').split(' ');
		return `type error: ${actualType} was given, but ${expectedType} is expected`;
	},
	UNREACHABLE_CODE: report => 'unreachable code',
	UNREACHABLE_STATE: report => `state ${report.info} is unreachable`,
};

function humanReportMessage(report: Report): string {
	const lines = [REPORT_MESSAGE_MAP[report.kind](report)];

	if (report.unexpectedTokenKind) {
		lines.push(`unexpected ${report.unexpectedTokenKind}`);
	}

	if (report.expectedTokenKinds.length > 0) {
		lines.push(`expected ${joinWithLast(report.expectedTokenKinds, ', ', ' or ')}`);
	}

	return lines.map((line, i) => (i > 0 ? ' - ' + line : line)).join('\n');
}

const MARKER_SEVERITY_MAP: Record<Report['severity'], monaco.MarkerSeverity> = {
	INFO: monaco.MarkerSeverity.Info,
	WARNING: monaco.MarkerSeverity.Warning,
	ERROR: monaco.MarkerSeverity.Error,
};

export function reportToMarkerData(report: Report): monaco.editor.IMarkerData {
	return {
		startLineNumber: report.selection.start.line,
		startColumn: report.selection.start.column,
		endLineNumber: report.selection.end.line,
		endColumn: report.selection.end.column + 1,

		severity: MARKER_SEVERITY_MAP[report.severity],

		message: humanReportMessage(report),
	};
}
